contributors/design-proposals/node/user-capabilities.md

```diff
@@ -0,0 +1,321 @@
+
+Most of the code is plumbing to expose this new field in the configs, then pass
+it to the Runtime through the CRI, then in the Runtime implementations, using
+it to populate the "ambient" capabilities.
+
+### Kubernetes
+
+An early WIP for adding the field to Kubernetes can be found
+[here](https://github.com/filbranden/kubernetes/commit/e8561087343c81478221a4cd6f8a9cc7e17cf502).
+It still needs more checking around the values that can be set here (as it's
+currently done for the root capabilities.)
+
+### containerd/cri
+
+For containerd/cri, a first step is to update to latest
+opencontainers/runtime-tools which gives more granular access to setting each
+capability set individually. PR containerd/cri#820 covers that.
+
+A follow up is to take the newly passed user capabilities and use them to
+populate the "ambient" capabilities, which can be done
+[here](https://github.com/containerd/cri/blob/v1.0.3/pkg/server/container_create.go#L375).
+
+### CRI-O
+
+Similar to containerd, "ambient" capabilities are currently being cleared in
+CRI-O, so it's also clear [where in the
+code](https://github.com/kubernetes-incubator/cri-o/blob/v1.10.3/server/container_create.go#L502)
+we should populate those using the new "userCapabilities".
+
+### User Interface
+
+The user will have access to this new feature through a new field in their
+"securityContext:" config for a container.
+
+As an example of a container running netcat as user "nobody" with
+`CAP_NET_BIND_SERVICE` to be able to bind to port 80, while also adding
+`CAP_SYS_NICE` to root inside container (in case someone uses `kubectl exec` on
+it to renice a process):
+
+```yaml
+apiVersion: v1
+kind: Pod
+spec:
+  containers:
+  - name: netcat
+    image: alpine
+    args:
+    - /bin/sh
+    - -c
+    - "nc -lk -p 80 -e echo hello"
+    securityContext:
+      runAsUser: nobody
+      userCapabilities:
+        add:
+        - NET_BIND_SERVICE
+      capabilities:
+        add:
+        - CAP_SYS_NICE
+```
+
+Note that adding "all" or "ALL" to "userCapabilities:" will have no effect,
+since doing otherwise would just amount to turning non-root inside container
+into real root. That's too dangerous, so let's just avoid it. Technically, it's
+possible to get a non-root user get all capabilities by listing them all
+explicitly, but at least that's explicit, so it's easier to catch.
+
+### Compatibility
+
+With the fields being passed as protobufs using gRPC, the absence of support for
+the field on either side just reverts to the default behavior, which is ignoring
+that user capabilities exist and keeping the "ambient" capabilities set to none.
+
+In effect, if either side lacks support, user capabilities will simply be
+silently dropped.
+
+This is the most secure setting and should not be too hard to troubleshoot given
+the failure scenario is likely to make the container fail quickly with a message
+that is likely to point to the lack of specific capabilities.
```

Comment:
It may be worth noting it will also just fail if the machine is running an older kernel (I can't remember off hand if it fails silently in `runc` though).